extension/dynamicconfig/service/remoteconfig.go

```diff
@@ -0,0 +1,156 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+
+	res "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+	pb "github.com/vmingchen/opentelemetry-proto/gen/go/collector/dynamicconfig/v1"
+)
+
+type UpdateStrategy uint8
+
+const (
+	Default UpdateStrategy = iota
+	OnGetFingerprint
+)
+
+type RemoteConfigBackend struct {
+	target		   string
+	conn           *grpc.ClientConn
+	client         pb.DynamicConfigClient
+	updateStrategy UpdateStrategy
+	chs            *responseMonitorChan
+}
+
+type responseMonitorChan struct {
+	getResp    chan *pb.ConfigResponse
+	updateResp chan *pb.ConfigResponse
+	quit       chan struct{}
+}
+
+func monitorResponse(chs *responseMonitorChan) {
+	var resp *pb.ConfigResponse
+
+	for {
+		select {
+		case chs.getResp <- resp:
+		case resp = <-chs.updateResp:
+		case <-chs.quit:
+			return
+		}
+	}
+}
+
+func NewRemoteConfigBackend(target string) (*RemoteConfigBackend, error) {
+	backend := &RemoteConfigBackend{
+		target: 		target,
+		conn:           nil,
+		client:         nil,
+		updateStrategy: Default,
+		chs: &responseMonitorChan{
+			getResp:    make(chan *pb.ConfigResponse),
+			updateResp: make(chan *pb.ConfigResponse),
+			quit:       make(chan struct{}),
+		},
+	}
+
+	if err := backend.initConn(); err != nil {
+		return nil, err
+	}
+
+	go monitorResponse(backend.chs)
+	return backend, nil
+}
+
+func (backend *RemoteConfigBackend) initConn() error {
+	conn, err := grpc.Dial(
+		backend.target,
+		grpc.WithInsecure(),
+	)
+	if err != nil {
+		return fmt.Errorf("remote config backend fail to connect: %w", err)
+	}
+
+	backend.conn = conn
+	backend.client = pb.NewDynamicConfigClient(conn)
+	return nil
+}
+
+
+func (backend *RemoteConfigBackend) GetUpdateStrategy() UpdateStrategy {
+	return backend.updateStrategy
+}
+
+func (backend *RemoteConfigBackend) SetUpdateStrategy(strategy UpdateStrategy) {
+	if strategy == Default || strategy == OnGetFingerprint {
+		backend.updateStrategy = strategy
+	}
+}
+
+func (backend *RemoteConfigBackend) GetFingerprint(resource *res.Resource) ([]byte, error) {
```

Comment:
Each call of GetFingerprint triggers a call to remote server. So the caller is response for not calling this too frequently and call it according to the server's suggested next call time? If so, please add comments.

Reply:
Yup that's correct, though as mentioned before, it may be a good idea to develop a more clever way of updating in the future. Comment added